### ISSUE TITLE
matchcoin (this app would do good to have its own crypto and distribution)

### ISSUE THREAD

**Issue by luvemanates** [luvemanates/matchthat#8]
Like the title says - matching should have its own currency because matching amounts has its own value.  It might also good to have a cutoff point (e.g.) this can only be matched up to a certain amount.

**Comment by luvemanates** [comment]
Need more details of how a separate coin is going to work with a charity based currency.  It needs to be somewhat centrailzed, and I'm thinking that instead of mining, it uses the web app to take note of certain matches to increase or decrease the value.

**Comment by luvemanates** [comment]
I'm having a problem understanding how the values would change over time.  e.g. Someone new matches for a radio benefit -- and then pledge $3000 then a new matchcoin should be created.  But as a currency value it might not change such as if someone decides to stop pledging a match.  It seems the value of a matchcoin would go up as new coins are created.  But in fact if someone takes a way a match it could still go up because the value of a match is a bit ghostly.

**Comment by luvemanates** [comment]
So I'm thinking that when a match is paid - you are given new matchcoins that have a base value.  But as you pay the value of your coin goes up.  But how does this coincide with the total value of everyone else's value?

**Comment by luvemanates** [comment]
maybe match coin needs a separate bank that assesses its own assets.  So for example, someone has been matching for years and has lots of coins.  As for the match they may have paid out millions of dollars, but the value of the matchcoin might still be 0.  So what inherently gives the matchcoin real value.  What actually gives the matchcoin value because it is ghostly.  The mechanism for giving coins will perhaps have to have an adjusted value as time goes on.  The value of assets change and more needs to be known about the charity based currency amount -- which this app would have to look at for what its worth where CBC is valued different based on different products.

**Comment by luvemanates** [comment]
It may go without saying that giving out a coin for instance 1 coin per 1 charity based currency coin might be standard.  But then as time goes on perhaps the value of those can multiply because the simple act of giving to charity creates more value as time goes on.

**Comment by luvemanates** [comment]
Also it seems that charity based currency, and urban development and basic necessity insurance would help provide a baseline that would also help value a match coin

**Comment by luvemanates** [comment]
I think both matches and tickets should be moderated which will help decipher a positive valuation

**Comment by luvemanates** [comment]
perhaps matchoin goes up in value for everyone with it the more matches reach the total amount matched to date.

**Comment by luvemanates** [comment]
The wallet should have a private key for initiating transfers with the rest of the system.  Optionally it can be stored on the server for ease of use or backup.  
But it needs to verify with the ledger on the server that this currency is contained in that account

**Comment by luvemanates** [comment]
Right now the mint is time based "1 coin per second" which is interesting because this way "time is money."  And the wallet for the bank is encrypted and shares a secret to succed with the mint wallet.  I was thinking the ledger might be better stored in DB rather than the current print to screen.  Either way its going to at least need a file if not a DB.

**Comment by luvemanates** [comment]
So, I was thinking about how to make this distributed, and here is some of the ruby code that would be neccessary to do that. 

https://reintech.io/blog/working-with-network-protocols-in-ruby

**Comment by luvemanates** [comment]
also, each coin should have a UUID, so the wallet has verification of a unique coin that has been deposited.

**Comment by luvemanates** [comment]
see https://www.rubydoc.info/gems/uuid/2.3.8/UUID
for uuid gen

**Comment by luvemanates** [comment]
added UUID, a ledger, and ledger entries.  Waiting to commit to get that new idea integrated

**Comment by luvemanates** [comment]
Okay, so the value baseline would be time, but as others contribute match through a charity based currency - that would create more value.  

e..g years have gone by and there are millions of matchcoin.  But as different programs get matched that bank coin is deposited in wallets across the world.  perhaps it becomes worth more, or less charity based currency.

**Comment by luvemanates** [comment]
Right now I'm trying to work out the details of how money is "owned" and how that money that is owned is transacted.

**Comment by luvemanates** [comment]
Currently the keys are stored in mongo, but they are in cleartext.  They need to be ciphered in and out of the database, and then deciphered upon being transferred to a new owner.  Also, the wallet needs to be password encrypted, and contain the secret hash to open the wallet's coin keys.

Also, it would be good to do a time based printing.  Currently doing one coin takes longer than a second, because the keys are being created too.  

As for distributing this I need to think about this.   It would be good to have some work -- right now I think it's the website that earns wallet's coin amounts.  But I think it would be better to stop using mysql and use Mongo for the whole system.

**Comment by luvemanates** [comment]
Also, all the debug printing should be moved to using a logger with level debug.

**Comment by luvemanates** [comment]
I'm thinking it might be good to have a client that is a bit like a punch card.  So the time you spent "doing something" is kept track of and you earn coins that way.  I think with match it might be good to have something where you can calculate how much something took in terms of time, and also the money that it represents.

**Comment by luvemanates** [comment]
The coin keys, and the wallet keys are now stored in the FS via mongodb.  Next we need to add passwords and hashing to wallets.  Also, indexes need to be added.

**Comment by luvemanates** [comment]
I'm thinking we could distribute this kind of coin by having someone make the key pair and send it to a tier and in return they are given a stake in the mint.  But with this wallets would be insecure

**Comment by luvemanates** [comment]
Currently there is an issue with creating change.  What can be done is if a specific is amount we make sure there is enough in the wallet, and then break up the 1 matchcoin amount into seperate serial numbers, with seperate keys.  

What does it do to take a private / public key and then hash it and store it on the ledger? This proves the ownership of a coin, or digital money that can be stored privately or publicly.  Public would be necessary to be distributed.

**Comment by luvemanates** [comment]
I think its imperative that private keys of a coin could be analyzed by a server, and then after that analysis, the keys could be changed -- and the amount could be also verified with a hash on the blockchain.

**Comment by luvemanates** [comment]
this coin needs a way to make change which is simply done with the following

Ask user what amount is needed, issue the amount needed after crediting from the wallet, and return the portion needed in one set of keys.  WIth the change as another set of keys.
SO there would be two sets of keys.  One for the amount, and one for the left over.  With the left over being returned to the wallet, and the amount to be paid, transferred. 

In other words we could Collect a bunch of 1 coins and create new face value coin with the bunch, and make new coins from the old keys.  Depending on the security the old keys may need to be burned e.g. the bank and the mint can re-use the old keys, but client wallets coin keys would have to be deleted.

**Comment by luvemanates** [comment]
to secure the keys with the matchcoin we could encrypt the serial number and verify it is UNIQUE to the blockchain.

**Comment by luvemanates** [comment]
So, now that we have merkle roots it would be good to keep coins based on the blockchain.  So in a transfer it would go as follows:
1) Sender authorizes coins to be sent to wallet address that is known on the blockchain.  This is done by using the sender wallet private key to sign a blockchain entry for a credit to send money to the receiver.
2)  Receiver verifies that the sender sent authorized the coin with the publicly viewable sender public key.  The receiver verifies they own the wallet by signing the block chain with the wallet's private key.  The node doing the computation verifieds that the receiver has signed ownership over the new coin through the wallet's public viewable key.  This updates the balances, and now the money has been transferred.
